### ISSUE TITLE
Crash on Linux after libX11 update with: [xcb] Unknown sequence number while processing queue (libX11 1.8.3 or 1.8.2 + X-IfEvent patch)

### ISSUE THREAD

**Issue by protodrew** [godotengine/godot#69352]
___

***Bugsquad note:** This issue has been confirmed several times already. No need to confirm it further.*

___

### Godot version

v4.0.beta6.official.7f8ecffa5

### System information

OpenSuse, Intel Iris XE, Wayland, Gnome

### Issue description

Trying to launch the godot beta 6 and it hangs upon opening menu or crashes outright.
```
Godot Engine v4.0.beta6.official.7f8ecffa5 - https://godotengine.org
libdbus-1.so: cannot open shared object file: No such file or directory
libdbus-1.so: cannot open shared object file: No such file or directory
OpenGL Renderer: Mesa Intel(R) Xe Graphics (TGL GT2)
 
[xcb] Unknown sequence number while processing queue
[xcb] You called XInitThreads, this is not your fault
[xcb] Aborting, sorry about that.
Godot_4_b6.x86_64: xcb_io.c:278: poll_for_event: Assertion `!xcb_xlib_threads_sequence_lost' failed.

```

### Steps to reproduce

Attempt to launch Godot 4 beta 6 on opensuse tumbleweed with wayland and gnome. I am using a framework laptop with 11th gen i5, and latest drivers for everything

### Minimal reproduction project

n/a

**Comment by clayjohn** [comment]
@Riteo May have some insight here. It looks like maybe your system can't fall back to X11 which is the only windowing library we support on linuxbsd right now. Riteo is in the process of adding wayland support and making the whole experience better for both X11 and Wayland

**Comment by Liemaeu** [comment]
Same issue here, openSUSE Tumbleweed, X11, KDE Plasma, Nvidia 525.60.11 driver.
Didn't have this issue with Beta 6 in the first place, only after updating my os yesterday.

`[xcb] Unknown sequence number while processing queue
[xcb] You called XInitThreads, this is not your fault
[xcb] Aborting, sorry about that.
Godot_v4.0-beta6_linux.x86_64: xcb_io.c:278: poll_for_event: Assertion `!xcb_xlib_threads_sequence_lost' failed.
`

**Comment by Liemaeu** [comment]
I recovered the snapshot from before I updated my os, works again.

Strangely only Godot 4 was affected (tried Beta 5, too), Godot 3.5 kept working.

**Comment by aaschmitz** [comment]
Same issue for me, OpenSUSE Tumbleweed, X11, KDE Plasma 5.26.3, Nvidia 525.60.11 driver, but using Godot 3.5.1.

The editor sometimes opens and crashes/closes after a few minutes, sometimes it doesn't even open the editor, other times it crashes/closes when running the game.

The issue started after the libX11-xcb1 version 1.8.2.x update, /var/log/messages show the entry:

`kwin_core: XCB error: 152 (BadDamage), sequence: 31178, resource id: 10499075, major code: 143 (DAMAGE), minor code: 3 (Subtract)`

Tried the Godot 3.5.1 from Official Site, the Steam Version, running it via Wine, all have the same issue. The only version that works now is Godot Flatpak.

**Comment by akien-mga** [comment]
Please report the bug to OpenSUSE and provide the information about which package caused the regression, so they can assign the package maintainer to look into it.

I have libX11 1.8.2 on Mageia 9 and no such problem. OpenSUSE seems to have a custom patch they added to fixed apparent regressions in Firefox, and this seems to have broken Godot: https://build.opensuse.org/package/show/openSUSE%3AFactory/libX11

**Comment by aaschmitz** [comment]
@Riteo Using several programs on X11 and I haven't noticed any other issues, it looks like only Godot is affected by this update to OpenSuse Tumbleweed.

**Comment by ionenwks** [comment]
Started hitting the same thing on Gentoo, traced it back to this libX11 patch which seems also used by OpenSUSE, not that anything but godot is giving me issues. Unclear if the problem is the patch or misuse somewhere that happened to work by accident.

https://github.com/gentoo/gentoo/blob/eeb3b760c603/x11-libs/libX11/files/libX11-1.8.2-reentrancy.patch
https://build.opensuse.org/package/view_file/openSUSE:Factory/libX11/U_Fix-797755-Allow-X-IfEvent-to-reenter-libX11.patch

**Comment by ionenwks** [comment]
https://gitlab.freedesktop.org/xorg/lib/libx11/-/issues/170

**Comment by llyyr** [comment]
Unrelated, but that patch also makes the same crash on gamescope

**Comment by bubbleguuum** [comment]
And that same patch crashes vkQuake.

**Comment by akien-mga** [comment]
We can keep it open so that affected users can find it. But yes it's something that needs to be fixed in libX11 / distro packages.

**Comment by llyyr** [comment]
opensuse have already fixed it as of 5 hours ago https://build.opensuse.org/package/show/openSUSE:Factory/libX11

**Comment by bubbleguuum** [comment]
It's not fixed as the patch is still in the  .spec file:
U_Fix-797755-Allow-X-IfEvent-to-reenter-libX11.patch

Also see this issue for further discussion in this hairy situation:
http://bugzilla.opensuse.org/show_bug.cgi?id=1205818

The end of this thread is about vkQuake, but this is the same issue affecting Godot.

**Comment by llyyr** [comment]
> It's not fixed as the problematic patch is still in the .spec file: U_Fix-797755-Allow-X-IfEvent-to-reenter-libX11.patch

no, it's fixed. the patch is for thread safety constructor which is disabled at build time now. Even if the patch is applied the file isn't built anyway

**Comment by bubbleguuum** [comment]
Except that according to my test disabling the safety constructor () does not fix this crash at all. At least on vkquake.
That's on a fresh libX11 1.8.2 compile with the problematic patch applied (https://gitlab.freedesktop.org/xorg/lib/libx11/-/commit/a9e845809bcaae22496bc8aa3ca252b410d5f39b) and `configure` option `--disable-thread-safety-constructor`.
With stock v1.8.2, it does not crash.

**Comment by ionenwks** [comment]
Yeah I gave it a quick try on Gentoo and the --disable option doesn't help godot if the patch is still applied, removing the patch or using libX11-1.8.1 works fine (simply removing would bring back firefox crashes though, maybe the --disable option still helps with that but not something I looked into).

**Comment by adventureronaquest** [comment]
Issue seems to be resolved with upgrade to
openSUSE Tumbleweed  20221205-0 -> 20221206-0

**Comment by bubbleguuum** [comment]
It reverted to 1.8.1 + some patches but without the problematic patch, so yes it does not crash anymore:

```
- Update to version 1.8.1
  This release fixes the --enable-thread-safety-constructor option to the
  configure script to work as intended.  In the previous release, the changes
  for this option may not have been enabled when the option was not specified
  or when the --enable option was specified.
  While we have enabled it by default, believing that doing so will reduce
  the number of bugs users encounter running libX11 clients, in some cases
  it may expose bugs in which clients had previously gotten away with calling
  libX11 functions while a libX11 lock is already held, and thus now deadlock,
  as discussed in https://gitlab.freedesktop.org/xorg/lib/libx11/-/issues/157  
- let's hope this version doesn't suffer yet from the regressions
  reported in boo#1205778, boo#1205818 (reported against 1.8.2);
  we need libX11 thread safe for totem (GNOME 43) :-(
```

**Comment by bubbleguuum** [comment]
Either that, or a bug on the problematic libX11 commit. Or in the very worst case, both.
Note sure I would agree with a patch in libX11 breaking (crashing) existing programs is the right thing to do, even if it expose some problem in said programs.

**Comment by RawDev1024** [comment]
> Issue seems to be resolved with upgrade to openSUSE Tumbleweed 20221205-0 -> 20221206-0

yes, upgrade libX11 from 1.8.1 to 1.8.2 is useful, my 3.5.1 stable and 4.0 beta 7 works again now.

**Comment by Zireael07** [comment]
Assuming 'note sure' is a typo for 'not sure'?

**Comment by aaschmitz** [comment]
With OpenSUSE Tumbleweed 20221206-0, the packages below are downgraded:

`libX11-6 libX11-6-32bit libX11-data libX11-xcb1 libX11-xcb1-32bit`

Now the Godot 3.5.1 are working fine again :+1:

**Comment by ionenwks** [comment]
ftr patch was a backport, upstream bug didn't get attention so it's now included in the now-released vanilla libX11-1.8.3 which doesn't seem to play well with godot still from a quick try

**Comment by Khhs167** [comment]
This issue has been reported in numerous other places, and seems to have been fixed just an hour ago as of writing this

https://github.com/ValveSoftware/Dota-2/issues/2218
https://gitlab.freedesktop.org/xorg/lib/libx11/-/merge_requests/176
https://bugs.archlinux.org/task/76860#comment213715

**Comment by rafael-arreola** [comment]
Fixed [Archlinux], I had to install downgrade with: 

```$ paru -Syu --noconfirm downgrade```

```$ sudo downgrade libx11 ```

  ->  and select the version libx11-1.8.2-2

**Comment by developomp** [comment]
Alternatively, you can also run

```
sudo pacman -U https://archive.archlinux.org/packages/l/libx11/libx11-1.8.2-2-x86_64.pkg.tar.zst
```

**Comment by underdoeg** [comment]
Looks like on arch it is already fixed

**Comment by developomp** [comment]
> Looks like on arch it is already fixed

Yup. Tested `Godot_v4.0-beta8_mono_linux_x86_64` with `libx11 1.8.3-3` and the issue seems to be fixed.

**Comment by llyyr** [comment]
It's "fixed" by distros (Arch) patching in PRs that aren't merged upstream yet (and might not even get merged) because they make crashes elsewhere, so this issue should probably stay open until it works without every distro rolling up their own patches

**Comment by akien-mga** [comment]
Yeah we should wait for an official upstream fix to be consistently deployed to all distros before closing. Despite it being open we had 5 duplicate bug reports, 2 of them even after we've pinned this issue.

But there's nothing to do from our side, this is just open for user information.

**Comment by Megalomaniak** [comment]
I don't see anyone mentioning XWayland in here, make sure you have it installed. That's what allows for X11 applications to run in wayland environments.

**Comment by cg9999** [comment]
I'm still randomly getting crashes during editor use on fully updated arch linux and godot 4.0 beta 10 / current master dev build.

```
> bin/godot.linuxbsd.editor.dev.x86_64
Godot Engine v4.0.beta.custom_build.f05fe7454 - https://godotengine.org
OpenGL Renderer: Mesa Intel(R) HD Graphics 620 (KBL GT2)
 
Editing project: /home/...[removed]
Godot Engine v4.0.beta.custom_build.f05fe7454 - https://godotengine.org
Vulkan API 1.3.230 - Using Vulkan Device #0: Intel - Intel(R) HD Graphics 620 (KBL GT2)
```

later various types of crashes with xcb:
```
[xcb] Unknown request in queue while dequeuing
[xcb] You called XInitThreads, this is not your fault
[xcb] Aborting, sorry about that.
godot.linuxbsd.editor.dev.x86_64: xcb_io.c:175: dequeue_pending_request: Assertion `!xcb_xlib_unknown_req_in_deq' failed.
```
or
```
godot.linuxbsd.editor.dev.x86_64: xcb_in.c:757: xcb_request_check: Assertion `!reply' failed.
```
or
```
[xcb] Unknown sequence number while processing queue
[xcb] You called XInitThreads, this is not your fault
[xcb] Aborting, sorry about that.
godot.linuxbsd.editor.dev.x86_64: xcb_io.c:278: poll_for_event: Assertion `!xcb_xlib_threads_sequence_lost' failed.
```

is this the same issue or something related. I haven't yet tried downgrading libx11 (It's currently arch 1.8.3-3, which should be already fixed)

**Comment by llyyr** [comment]
Downgrade to libx11 1.7.5, or build it with `--disable-thread-safety-constructor`

**Comment by Eisendroid** [comment]
It seems like it's still broken.  I am on EndeavourOS (Arch) and it got fixed by some update few weeks ago (downgrade?). Today libx11 got pushed to 1.8.3-4 and it's broken again.

**Comment by akien-mga** [comment]
Arch Linux users, please comment on https://bugs.archlinux.org/task/76860 which is the upstream report.

1.8.3-4 removed the (yet unmerged) upstream patches that solve the issue, to replace them with a (merged) upstream patch. They seem to expect it fixes Godot too, but evidently it doesn't. https://github.com/archlinux/svntogit-packages/commit/27567a0220da86d3c45eb3e341f30457f93d99d7

**Comment by Khhs167** [comment]
Still experiencing crashes in godot 4 beta 10 on 1.8.3-5 on arch. So bug tracker issue is reopened over there

**Comment by cg9999** [comment]
> Still experiencing crashes in godot 4 beta 10 on 1.8.3-5 on arch. So bug tracker issue is reopened over there

Same here. -4 was completely broken. Either hang or crash at startup, but -5 is back to random crashing behaviour.

**Comment by Megalomaniak** [comment]
lib32-libx11 1.8.3-1.0 update that I was greeted with this morning on Manjaro Testing seems to have resolved it for me.

**Comment by LinuxUserGD** [comment]
Latest upstream version [ba548ed1](https://gitlab.freedesktop.org/xorg/lib/libx11/-/commit/ba548ed1688d4d5a83913218155d0ddf616ff176) (`libX11-9999.ebuild`) works fine on Gentoo
<details>

```bash
# Copyright 1999-2022 Gentoo Authors
# Distributed under the terms of the GNU General Public License v2

EAPI=7

XORG_DOC=doc
XORG_MULTILIB=yes
XORG_TARBALL_SUFFIX=xz
inherit toolchain-funcs xorg-3 git-r3

# Note: please bump this with x11-misc/compose-tables
DESCRIPTION="X.Org X11 library"
EGIT_REPO_URI="https://gitlab.freedesktop.org/xorg/lib/libx11"
KEYWORDS="~alpha ~amd64 ~arm ~arm64 ~hppa ~ia64 ~loong ~m68k ~mips ~ppc ~ppc64 ~riscv ~s390 ~sparc ~x86 ~x64-cygwin ~amd64-linux ~x86-linux ~ppc-macos ~x64-macos"
IUSE="test"
RESTRICT="!test? ( test )"

RDEPEND="
	>=x11-libs/libxcb-1.11.1[${MULTILIB_USEDEP}]
	x11-misc/compose-tables

	!<xfce-base/xfce4-settings-4.16.3
"
DEPEND="${RDEPEND}
	x11-base/xorg-proto
	x11-libs/xtrans
"
BDEPEND="test? ( dev-lang/perl )"

src_configure() {
	local XORG_CONFIGURE_OPTIONS=(
		$(use_with doc xmlto)
		$(use_enable doc specs)
		--enable-ipv6
		--without-fop
		CPP="$(tc-getPROG CPP cpp)"
	)
	xorg-3_src_configure
}

src_install() {
	xorg-3_src_install
	rm -rf "${ED}"/usr/share/X11/locale || die
}
```
</details>

**Comment by ionenwks** [comment]
> Latest upstream version [ba548ed1](https://gitlab.freedesktop.org/xorg/lib/libx11/-/commit/ba548ed1688d4d5a83913218155d0ddf616ff176) (`libX11-9999.ebuild`) works fine on Gentoo

Gentoo has backported fixes in libX11-1.8.3-r1 which afaik works fine.

**Comment by LinuxUserGD** [comment]
Ok, seems like they have been merged upstream too.

**Comment by akien-mga** [comment]
> lib32-libx11 1.8.3-1.0 update that I was greeted with this morning on Manjaro Testing seems to have resolved it for me.

I'm a bit confused by the versioning here, how does 1.8.3-1.0 compare to 1.8.3-5 from stock Arch repos?

The lib32-libx11 and libx11 packages seem weirdly out of sync:
![image](https://user-images.githubusercontent.com/4701338/213246077-8fd69f1a-d3e0-4332-9df0-c23888a58572.png)

How is lib32-libx11 used exactly? Isn't that only the 32-bit equivalent of libx11, and thus used only if you run a 32-bit build of Godot?

> Gentoo has backported fixes in libX11-1.8.3-r1 which afaik works fine.

That's interesting because this package backports the same two patches as Arch Linux's 1.8.3-5, yet users are reporting that this one still crashes intermittently. Please let us know if you do experience intermittent crashes eventually pointing at X11 stuff.

**Comment by ionenwks** [comment]
> > Gentoo has backported fixes in libX11-1.8.3-r1 which afaik works fine.
> 
> That's interesting because this package backports the same two patches as Arch Linux's 1.8.3-5, yet users are reporting that this one still crashes intermittently. Please let us know if you do experience intermittent crashes eventually pointing at X11 stuff.

Don't know if it's *really* fixed (I haven't tested enough to pickup intermittent crashes if they're rare), but Arch has an older version of the PR and isn't the same as what has been merged upstream or used by Gentoo.

e.g. 176.diff adds
```
+    if (dpy->in_ifevent && dpy->ifevent_thread == xthread_self())
+        return;
```
while libX11-1.8.3-reentrancy-again.patch instead does:
```
+    if (dpy->in_ifevent && xthread_equal(dpy->ifevent_thread, xthread_self()))
+        return;
```

**Comment by LinuxUserGD** [comment]
> 
> (**Edit:** not familiar enough with libX11 to know if these differences can have an impact for Godot, but if the equal tests fail I'd assume it'd be bad)

`176.diff` probably is an old version of the PR, see [#176](https://gitlab.freedesktop.org/xorg/lib/libx11/-/merge_requests/176/diffs?commit_id=4cffcfd1735f8c6411cc1aee00e6e92194929671#08a5218775e840303de6401e035a825b418a0ddd_473_464) while gentoo uses the backported version of the merged PR [#176](https://gitlab.freedesktop.org/xorg/lib/libx11/-/merge_requests/176/diffs#08a5218775e840303de6401e035a825b418a0ddd_473_464)

**Comment by akien-mga** [comment]
@ionenwks @LinuxUserGD Good hint! I notified the Arch Linux libx11 maintainer who re-synced the patches with upstream in 1.8.3-6.

Could Arch Linux users who were still experiencing intermittent crashes confirm that 1.8.3-6 solves it for you?

**Comment by Megalomaniak** [comment]
> I'm a bit confused by the versioning here, how does 1.8.3-1.0 compare to 1.8.3-5 from stock Arch repos?

It was the version string I had remembered, along it other libX11 libraries including the 64 bit ones were updated too.

**Comment by daviirodrig** [comment]
I can confirm that updating libx11 to version 1.8.3-6 on Arch Linux successfully resolved the issue.

**Comment by akien-mga** [comment]
Thanks everyone for the reports!

I believe the issue has been properly fixed in upstream libx11 now (for a future 1.8.4 release), and most Linux distros seem to be aware of the issues with 1.8.2 and 1.8.3 and have found ways to mitigate them:
- Stick to 1.8.1 for OpenSUSE
- Stick to 1.8.2 + patch for Fedora (1.8.3 without MR 176 was proposed as update candidate but is currently on hold due to those very issues: https://bodhi.fedoraproject.org/updates/FEDORA-2023-be3023012d )
- Update to 1.8.3 + upstream MR 176 and 180 (Arch, Gentoo, Mageia)

So let's close :)

**Comment by MitrB** [comment]
I'm crashing, even with version 1.8.4
```
[xcb] Unknown sequence number while awaiting reply
[xcb] You called XInitThreads, this is not your fault
[xcb] Aborting, sorry about that.
Godot_v4.0.2-stable_linux.x86_64: xcb_io.c:374: poll_for_response: Assertion `!xcb_xlib_threads_sequence_lost' failed.
```
I do have libx11 version 1.8.4.
```
Name            : libx11
Version         : 1.8.4-1
Description     : X11 client-side library
Architecture    : x86_64
URL             : https://xorg.freedesktop.org/
Licenses        : custom
Groups          : None
Provides        : None
Depends On      : libxcb  xorgproto
Optional Deps   : None
Required By     : aalib  at-spi2-core  cairo  dmenu  ffmpeg  ffmpeg4.4  gst-plugins-bad-libs  gtk3  gtk4
                  lib32-libx11  libva  libxext  libxfixes  libxkbfile  libxrender  libxt  lightdm
                  minecraft-launcher  mpv  pcsx2  sdl2  startup-notification  vulkan-intel  vulkan-tools
                  vulkan-validation-layers  webkit2gtk  webkit2gtk-4.1  xbindkeys  xorg-xdpyinfo  xorg-xinit
                  xorg-xinput  xorg-xkill  xorg-xmodmap  xorg-xprop  xorg-xrandr  xorg-xrdb  zvbi
Optional For    : glew  hwloc  mtools  suil
Conflicts With  : None
Replaces        : None
Installed Size  : 10.01 MiB
Packager        : Andreas Radke <andyrtr@archlinux.org>
Build Date      : Sat 04 Feb 2023 21:54:12 CET
Install Date    : Mon 13 Feb 2023 12:07:20 CET
Install Reason  : Explicitly installed
Install Script  : No
Validated By    : Signature

```

**Comment by cg9999** [comment]
> I'm crashing, even with version 1.8.4 . It's hard to replicate the crashes, as they seem random to me. Sometimes I can go hours without a crash, sometimes I have one shortly after starting Godot.

Seems to be related to Arch system. More discussion in  #75308

**Comment by isosphere** [comment]
I had this issue in Debian bookworm, which currently ships with libx11-6 version 1.8.4-2+deb12u2. 

I installed the `deb` [package from the testing release (trixie)](https://packages.debian.org/trixie/libx11-6), which is version 1.8.6-1. The installation of this package over the existing went smoothly and I believe it has resolved the issue for me - though I'm not entirely sure that my actions will have no unwanted side effects with other parts of my system.

**Comment by weidengeist** [comment]
It would be pretty interesting, if the issue was solved for you. I’m running libX11 1.8.6-1 (Arch Linux) as well and unfortunately I’m still experiencing those crashes. Please let us know if the problems still persist for you or not. Maybe it’s another component or the interaction of components that has to be taken into consideration to resolve those crashes once and for all.
